### PR TITLE
[ts-bodies] Support for fallthrough in switch

### PR DESCRIPTION
**Pull request by trilis**

### CODE REVIEW

File: typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt [hunk 1]
```diff
@@ -145,23 +150,103 @@ internal class ExpressionConverter(val documentConverter: DocumentConverter) {
...
+                }
+                val body = map { case ->
+                    if (case.condition == null) {
+                        case.body.statements.map { it.convert() }
+                    } else {
+                        listOf(
+                            IfStatementModel(
+                                BinaryExpressionModel(
+                                    expressionToCompare,
+                                    "==",
```
Comment (Schahen) [inline typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt:225]: We should finally narrow BinaryExpressionModel::operator but it wouldn't be honest from my side to ask you to do it in this particular PR.  However what makes sense  even in the framework of this PR is to move this operators to some constants. Can you please do it?

File: typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt [hunk 2]
```diff
@@ -145,23 +150,103 @@ internal class ExpressionConverter(val documentConverter: DocumentConverter) {
...
+
         return WhenStatementModel(
-            expression.convert(),
-            cases.map { CaseModel(
-                it.condition?.convert(),
-                convertBlock(it.body)
-            ) }
+            expressionToCompare,
+            splitToFallthroughBlocks()
+                .mapNotNull { it.convert(expressionToCompare) }
```
Comment (Schahen) [inline typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt:248]: why we need to maps here? won't one be enough?

File: typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt [hunk 3]
```diff
@@ -145,23 +150,103 @@ internal class ExpressionConverter(val documentConverter: DocumentConverter) {
...
+    private fun List<CaseDeclaration>.convert(expressionToCompare: ExpressionModel): CaseModel? {
+        return when (this.size) {
+            0 -> null
+            1 -> CaseModel(
+                first().condition?.convert()?.let { listOf(it) },
+                convertBlock(first().body)
+            )
+            else -> {
+                val condition = if (none { it.condition == null }) {
+                    map { it.condition!!.convert() }
```
Comment (Schahen) [inline typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt:213]: whenever it's possible not to to use `!!` we shouldn't use it.

File: typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt (hunk 1 above)
Comment (trilis) [reply typescript/ts-node-lowering/src/org/jetrbains/dukat/nodeLowering/lowerings/convertStatements.kt]: Fixed - now binary and unary operators are enums in kotlin model

**Comment by trilis** [comment]
43ba21c40a4a207f17a2d9bc59c1b2e28d0ab9ed

**Pull request closed without merging**